Do not enable venv in terminal for bash-like oneshot task invocations

WeetHet opened the pull request:
Release Notes:
- Work around #8334 by only activating venv in the terminal not in tasks (see #8440 for a proper solution)
- To use venv modify your tasks in the following way:
```json
{
  "label": "Python main.py",
  "command": "sh",
  "args": ["-c", "source .venv/bin/activate && python3 main.py"]
}
```

WeetHet committed:
Workaround #8334

WeetHet commented:
@cla-bot check

WeetHet commented:
@cla-bot check

SomeoneToIgnore reviewed (commented):
I wonder about different cases possible here:
* a regular terminal
* a generic static task from tasks.json, `bash -c ...`, without any python-specific knowledge
* a dynamic task, added for some *.py file from the worktree (worktree is, for example, a drag-and-dropped folder with some other Python project, or just a drag-and-dropped *.py file script)

Would some of them want to have venv configured?
E.g those dynamic tasks?

For the rest — is it so, that we do need some "static" venv that we get from `source .venv/bin/activate`? Could there be different ways that static thin be retrieved with? ($PATH variable contents?)
If so, we can try adding that variant implicitly to all "static" tasks now, that looks better if my assessments are correct.

WeetHet commented:
> I wonder about different cases possible here:
> 
>     * a regular terminal
> 
>     * a generic static task from tasks.json, `bash -c ...`, without any python-specific knowledge
> 
>     * a dynamic task, added for some *.py file from the worktree (worktree is, for example, a drag-and-dropped folder with some other Python project, or just a drag-and-dropped *.py file script)
> 
> 
> Would some of them want to have venv configured? E.g those dynamic tasks?
> 
> For the rest — is it so, that we do need some "static" venv that we get from `source .venv/bin/activate`? Could there be different ways that static thin be retrieved with? ($PATH variable contents?) If so, we can try adding that variant implicitly to all "static" tasks now, that looks better if my assessments are correct.

All of the solutions you've proposed somewhat resemble the solution I've proposed I'm #8440. They require more changes and implementing them right now would mean that Python tasks would stay broken for longer. This pull request is no more than a temporary workaround until a better solution gets implemented latter on

WeetHet committed:
Add venv/bin to path for tasks without using activate scripts

WeetHet commented:
@SomeoneToIgnore I have implemented a function to use venv inside the tasks and it seems to be working. The only thing is, I need to clean up the PATH after the task is completed and I don't know where it is. I'd appreciate your help on that

WeetHet committed:
Fix hashmap type

SomeoneToIgnore reviewed (commented):
> the only thing is, I need to clean up the PATH after the task is completed and I don't know where it is

I've just realized that, but we do set `env` values globally because that's what 
https://github.com/zed-industries/zed/blob/8b5a732671749704934522d147145f95b20f862c/crates/terminal/src/terminal.rs#L336-L344
does?? Looks dire.

I think what we can do now is to put the old env into `TaskState` (https://github.com/zed-industries/zed/blob/8b5a732671749704934522d147145f95b20f862c/crates/terminal/src/terminal.rs#L602) and react on the `::Exit` event here, cleaning up things:
https://github.com/zed-industries/zed/blob/8b5a732671749704934522d147145f95b20f862c/crates/terminal/src/terminal.rs#L639-L642

Much better approach would be to propose a change to Alacritty, that seem to semi-expose the process things already, here's the place I would imagine we need to change:
https://github.com/alacritty/alacritty/blob/37349c4566a1431b5becbb649dec536ecbe18e71/alacritty_terminal/src/tty/unix.rs#L228-L230

WeetHet commented:
> > the only thing is, I need to clean up the PATH after the task is completed and I don't know where it is
> 
> I've just realized that, but we do set `env` values globally because that's what
> 
> https://github.com/zed-industries/zed/blob/8b5a732671749704934522d147145f95b20f862c/crates/terminal/src/terminal.rs#L336-L344
> 
> 
> does?? Looks dire.
> I think what we can do now is to put the old env into `TaskState` (
> 
> https://github.com/zed-industries/zed/blob/8b5a732671749704934522d147145f95b20f862c/crates/terminal/src/terminal.rs#L602
> 
> ) and react on the `::Exit` event here, cleaning up things:
> https://github.com/zed-industries/zed/blob/8b5a732671749704934522d147145f95b20f862c/crates/terminal/src/terminal.rs#L639-L642
> 
> Much better approach would be to propose a change to Alacritty, that seem to semi-expose the process things already, here's the place I would imagine we need to change: https://github.com/alacritty/alacritty/blob/37349c4566a1431b5becbb649dec536ecbe18e71/alacritty_terminal/src/tty/unix.rs#L228-L230

Sure, yeah, feel free to do that

WeetHet commented on crates/project/src/terminals.rs:
```diff
@@ -28,10 +28,22 @@ impl Project {
         let mut env = settings.env.clone();
+
+        // if we are creating a task, activate minimal python virtual environment
+        if !is_terminal {
```
We can't move them both because the full venv activation should commence after the terminal is created and `set_python_venv_path_for_tasks` is done before. On the other hand, I did in fact overlook the `if let Some(spawn_task) = spawn_task {` branch, so I'd move the minimal venv activation there

WeetHet committed:
Move minimal venv activation into task creation

SomeoneToIgnore commented:
Let's do it the proper way and implement https://github.com/alacritty/alacritty/issues/7778 in Zed's fork at least (ideally, upstream it as it does not look like something hard) and return back to this PR, to make it proper.

Thank you a lot for pointing out the env flaw we have now.

SomeoneToIgnore commented:
Update: https://github.com/alacritty/alacritty/pull/7788 is working for the unix part and I have the branch with that integration hardcoded: https://github.com/zed-industries/zed/tree/kb/use-new-alacritty so let's see how the review goes.

WeetHet commented:
> Update: [alacritty/alacritty#7788](https://github.com/alacritty/alacritty/pull/7788) is working for the unix part and I have the branch with that integration hardcoded: https://github.com/zed-industries/zed/tree/kb/use-new-alacritty so let's see how the review goes.

That's nice. You probably should probably add my venv activation and test if it works as expected

WeetHet committed:
Use new alacritty api for child envs

WeetHet committed:
Fix formatting and clippy

WeetHet commented:
@SomeoneToIgnore I've added the code to use your new alacritty API. The only question now is: do we pin the alacritty_terminal revision now or wait until release?

SomeoneToIgnore commented on crates/project/src/terminals.rs:
```diff
@@ -128,6 +141,41 @@ impl Project {
+        working_directory: Option<PathBuf>,
+        env: &mut collections::HashMap<String, String>,
+    ) {
+        let working_directory = working_directory.unwrap_or_else(|| Path::new("/").to_path_buf());
```
Let's at least deal with the comment from
https://github.com/zed-industries/zed/blob/5139aa38117a36e9e55abdcb047b9eafc84c6f45/crates/project/src/terminals.rs#L107-L109

I also wonder, whether defaulting to `/`+arbitrary string from the config is a good default.
For alacritty, `working_directory` boils down to `Command::process` behavior of Rust stdlib:
https://github.com/alacritty/alacritty/blob/406170050d1b08a0ad717d9e6746c6369bbf262a/alacritty_terminal/src/tty/unix.rs#L251-L254

which defaults to relative path usage.
Could it also be more appropriate for python env case?

WeetHet replied:
That was not my decision to make, so I'd say that is intended behaviour.

WeetHet replied:
Sure, this works for me

SomeoneToIgnore commented on crates/project/src/terminals.rs:
```diff
@@ -128,6 +141,41 @@ impl Project {
+
+                let path_bin = path.join("bin");
+                // We need to set the PATH to include the virtual environment's bin directory
+                if let Some(paths) = std::env::var_os("PATH") {
```
A few notes I've got for this code block:

* we could replace `paths.insert(0, path_bin);` with `std::iter::once(path_bin).chain(paths...)` and avoid `collect` at all, since `join_paths` allows all [iterator-like](https://doc.rust-lang.org/std/env/fn.join_paths.html) arguments.

* env paths are the same every `for virtual_environment_name` entry, but we always check `PATH` variable, split it, filter it, collect it.
We could split it once into a HashMap that holds var + order number (if it matters for env vars?) and do `so we don't add the same path twice` checks more effectively.

* If the venv bin path is not contained in PATH, we could avoid re-joining the path elements entirely, rather prepending the `path_bin` before the unsplit env string.
Even more: we effectively run `.filter(|p| p != &path_bin)` on every `working_directory.join(virtual_environment_name)` which we can also deduplicate and pre-join upfront.

* would broken utf-8 path that we get with `to_string_lossy().to_string()` work in any of the OSes as a PATH variable? Not insisting, but we could omit, logging such cases?

WeetHet replied:
- Okay, done
- Not really, there is a `if path.exists()`, so the code executes only once
- Removed the filter, as the variables are now local, so there is no such problem anymore
- I don't know what to do about this, the best I could come up with is logging out if the path is broken, as env right now accepts String, not OsString

SomeoneToIgnore replied:
Now I'm totally confused by the `set_python_venv_path_for_tasks` and `find_activate_script_path` methods.

They both iterate over `settings.directories` (which is a horrible name with no docs, what are those directories are, in the first place since they are referenced as `virtual_environment_name`??) and check such dirs for existence.

* `find_activate_script_path` takes the first existing script and uses that for the activation
* `set_python_venv_path_for_tasks` instead, goes over every `virtual_environment_name` and takes the **last** entry that exists, setting it into the `env`

What's additionally confusing with the `set_python_venv_path_for_tasks` is the fact that it does std::env PATH var splitting and collecting only to discard it if there's another existing `virtual_environment_name` path in the list.

Ergo the questions:

* what does `set_python_venv_path_for_tasks` intend to do and why it takes different paths compared to `find_activate_script_path`?
* if `set_python_venv_path_for_tasks` is correctly written, should it do `.rev().filter(|path| path.exists()).next()` and do the `std::env` split-join dance once only?

WeetHet replied:
Oh, `set_python_venv_path_for_tasks` should just find the venv path and add it to the path, so, basically just `find(|path| path.exists()` and then split/join

WeetHet replied:
Okay, yeah, I've rewritten both of these functions to use iterators

SomeoneToIgnore reviewed (commented):
Thank you for the update.

We could use https://github.com/alacritty/alacritty/releases/tag/v0.13.2-rc1 and update the TODO above it a bit.

WeetHet committed:
Always set ZED_TERM to true

WeetHet commented on crates/terminal/src/terminal.rs:
```diff
@@ -650,6 +650,9 @@ impl Terminal {
                     .push_back(InternalEvent::ColorRequest(*idx, fun_ptr.clone()));
             }
+            AlacTermEvent::ChildExit(_) => {
+                // TODO: Handle child exit
```
Sure, that's just not the area this pull request covers

WeetHet committed:
Don't collect the paths

WeetHet commented on crates/terminal/src/terminal.rs:
```diff
@@ -310,13 +310,20 @@ impl TerminalBuilder {
+        env.entry("LC_ALL".to_string())
+            .or_insert_with(|| "en_US.UTF-8".to_string());
+
+        env.entry("ZED_TERM".to_string())
```
Done

WeetHet committed:
Reformat

SomeoneToIgnore committed:
Use fixed Alacritty version

SomeoneToIgnore committed:
No unwraps

SomeoneToIgnore committed:
Do not use `/` for venv base path when detecting venv

SomeoneToIgnore committed:
Remove nested `Some` check in the venv activation

WeetHet committed:
Iterator based path-search for better clarity

SomeoneToIgnore reviewed (approved):
Ok, there are still 3 things that might go wrong:

* `new_path.to_string_lossy().to_string()` thing that most probably makes no sense when put into PATH env var

But does not seem very bad overall and I think we need to deal with `python_settings.directories` side first to make things better here.

* calling `.exists()` directly instead of relying of `Fs`

We do not call any `std::fs` methods directly due to various reasons.
Here it does not matter much since we do not have any good tests for terminal, and all `Fs` methods are async so this won't be very pleasant to write.

*


```
let venv_base_directory = working_directory
            .as_deref()
            .or(current_directory.as_deref())
            .unwrap_or_else(|| Path::new("")); // if everything fails, use relative path
```

neither of this is a process' parent directory, so not sure we're doing thing right by using the `current_directory`.

Cases when we do not have a working directory are almost not existing now, so does not look critical, esp. given the fact that I failed to get any answer from you that makes sense on "what path does venv want".


----------------------

So looks like we're ready for the first iteration with this.

Feel free to follow-up any of these or the new TODO if you want to, and thank you again for all the env fixups and emphasis.

SomeoneToIgnore commented:
After all these changes, I'm not sure what's the status with all the related issues, so merging with the original commit message — would be great to get those issues updated with what's left to be done for them.

WeetHet commented:
Oh, wait, the part with current_directory should not have made it into

WeetHet commented:
@SomeoneToIgnore is there a way to unmerge? If no, could you hotfix this part in main? 
```rust
let venv_base_directory = working_directory
            .as_deref()
            .or(current_directory.as_deref())
            .unwrap_or_else(|| Path::new("")); // if everything fails, use relative path
```

Should be

```rust
let venv_base_directory = working_directory
    .as_deref()
    .unwrap_or_else(|| Path::new(""));
```

SomeoneToIgnore commented:
Please, create a follow-up PR (as mentioned in my latest comment here).

Would be super cool to ensure things that were improved in this PR are working for you as intended.

WeetHet commented:
> Please, create a follow-up PR (as mentioned in my latest comment here).
> 
> Would be super cool to ensure things that were improved in this PR are working for you as intended.

Okay, done in #9661
